### PR TITLE
feat: Add impossible_comparisons #76

### PR DESCRIPTION
**Pull request by MariangelaNM**
**Issue**
[76](https://github.com/keep-starknet-strange/cairo-lint/issues/76)

**Description**
1.  Implement is_impossible_double_comparison.
2. Fix the current implementation.
3. Add and update tests.

### CODE REVIEW

File: crates/cairo-lint-core/src/lints/double_comparison.rs [hunk 1]
```diff
@@ -6,6 +6,8 @@ use cairo_lang_syntax::node::ids::SyntaxStablePtrId;
 use cairo_lang_syntax::node::{TypedStablePtr, TypedSyntaxNode};
 
 pub const SIMPLIFIABLE_COMPARISON: &str = "This double comparison can be simplified.";
+pub const IMPOSSIBLE_COMPARISON: &str =
+    "The condition contains contradictory comparisons, meaning no value can satisfy both conditions at once.";
```
Comment (0xLucqs) [inline crates/cairo-lint-core/src/lints/double_comparison.rs:10]: this is redundant with `pub const CONTRADICTORY_COMPARISON: &str = "This double comparison is contradictory and always false.";`

**Comment by MariangelaNM** [comment]
@0xLucqs

File: crates/cairo-lint-core/src/lints/double_comparison.rs [hunk 2]
```diff
@@ -35,6 +37,12 @@ pub fn check_double_comparison(
                     binary_expr.stable_ptr().untyped(),
                     Severity::Warning,
                 ));
+            } else if is_impossible_double_comparison(&lhs_op, &rhs_op, &middle_op) {
+                diagnostics.push(create_diagnostic(
+                    IMPOSSIBLE_COMPARISON,
+                    binary_expr.stable_ptr().untyped(),
+                    Severity::Error,
+                ));
```
Comment (0xLucqs) [inline crates/cairo-lint-core/src/lints/double_comparison.rs:45]: this is redundant with `is_contradictory_double_comparison` can you just add your code to this function ? or even better move this into a new lint

File: crates/cairo-lint-core/tests/test_files/double_comparison/double_comparison [hunk 3]
```diff
@@ -399,3 +399,37 @@ fn main() -> bool {
         false
     }
 }
+
```
Comment (0xLucqs) [inline crates/cairo-lint-core/tests/test_files/double_comparison/double_comparison:402]: add 1 test for every impossible case

**Review by 0xLucqs — CHANGES_REQUESTED** [review]

**Comment by MariangelaNM** [comment]
@0xLucqs

**Comment by MariangelaNM** [comment]
@0xLucqs , I joined the two logics in is_impossible_double_comparison and added two more examples.

File: crates/cairo-lint-core/src/lints/double_comparison.rs [hunk 4]
```diff
@@ -97,20 +97,59 @@ fn is_redundant_double_comparison(
...
+
+fn extract_value_from_operator(op: &BinaryOperator) -> Option<i32> {
+    match op {
+        BinaryOperator::EqEq(_)
+        | BinaryOperator::LT(_)
+        | BinaryOperator::LE(_)
+        | BinaryOperator::GT(_)
+        | BinaryOperator::GE(_) => Some(0), 
+        _ => None,
+    }
```
Comment (0xLucqs) [inline crates/cairo-lint-core/src/lints/double_comparison.rs:152]: I don't understand what you're doing here

**Comment by MariangelaNM** [comment]
@0xLucqs ,I updated the initial function to include two additional cases: checking for `x > y && x >= y` and `x <= y && x > y`.

File: crates/cairo-lint-core/src/lints/double_comparison.rs [hunk 5]
```diff
@@ -110,9 +111,12 @@ fn is_contradictory_double_comparison(
             | (BinaryOperator::GT(_), BinaryOperator::AndAnd(_), BinaryOperator::EqEq(_))
             | (BinaryOperator::LT(_), BinaryOperator::AndAnd(_), BinaryOperator::GT(_))
             | (BinaryOperator::GT(_), BinaryOperator::AndAnd(_), BinaryOperator::LT(_))
+            | (BinaryOperator::GT(_), BinaryOperator::AndAnd(_), BinaryOperator::GE(_))
+            | (BinaryOperator::LE(_), BinaryOperator::AndAnd(_), BinaryOperator::GT(_))
```
Comment (0xLucqs) [inline crates/cairo-lint-core/src/lints/double_comparison.rs:115]: these are not contradictory

Comment (0xLucqs) [reply crates/cairo-lint-core/src/lints/double_comparison.rs:115]: they're redundant

**Comment by MariangelaNM** [comment]
@0xLucqs ,Based on your feedback, I adjusted the logic and placed it in `is_redundant_double_comparison`

**Review by 0xLucqs — APPROVED** [review]

**Comment by 0xLucqs** [comment]
@all-contributors please add @MariangelaNM for code and tests

**Pull request merged**